src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs

```diff
@@ -218,18 +231,46 @@ private IEnumerable<ProjectCacheDescriptor> GetProjectCacheDescriptors(ProjectIn
                     ProjectCacheException.ThrowForErrorLoggedInsideTheProjectCache("ProjectCacheInitializationFailed");
                 }
 
-                return new ProjectCachePlugin(pluginTypeName, pluginInstance);
+                FileAccessManager.HandlerRegistration handlerRegistration = _fileAccessManager.RegisterHandlers(
+                    (buildRequest, fileAccessData) =>
+                    {
+                        // TODO: Filter out projects which do not configure this plugin
+                        FileAccessContext fileAccessContext = GetFileAccessContext(buildRequest);
+                        pluginInstance.HandleFileAccess(fileAccessContext, fileAccessData);
+                    },
+                    (buildRequest, processData) =>
+                    {
+                        // TODO: Filter out projects which do not configure this plugin
+                        FileAccessContext fileAccessContext = GetFileAccessContext(buildRequest);
+                        pluginInstance.HandleProcess(fileAccessContext, processData);
+                    });
+
+                return new ProjectCachePlugin(pluginTypeName, pluginInstance, handlerRegistration);
             }
             catch (Exception e)
             {
-                return new ProjectCachePlugin(pluginTypeName, Instance: null, ExceptionDispatchInfo.Capture(e));
+                return new ProjectCachePlugin(pluginTypeName, Instance: null, HandlerRegistration: null, ExceptionDispatchInfo.Capture(e));
             }
             finally
             {
                 MSBuildEventSource.Log.ProjectCacheBeginBuildStop(pluginTypeName);
             }
         }
 
+        private FileAccessContext GetFileAccessContext(BuildRequest buildRequest)
+        {
+            BuildRequestConfiguration configuration = _configCache[buildRequest.ConfigurationId];
+
+            // TODO dfederm: This allocates a lot.
+            var globalProperties = new Dictionary<string, string>(configuration.GlobalProperties.Count, StringComparer.OrdinalIgnoreCase);
+            foreach (ProjectPropertyInstance property in configuration.GlobalProperties)
+            {
+                globalProperties.Add(property.Name, property.EvaluatedValue);
+            }
+
+            return new FileAccessContext(configuration.ProjectFullPath, globalProperties, buildRequest.Targets);
```

Comment:
Just curious - why are the gloval properties of concern at all?
I'd naively think that for deterministic build only the initial state would be of concern (plus the list of all the files participating in the build). Or is it so that we can skip portions (targets?) of the project build, while still building others? If yes - is there a granularity unit (target?) for which we can group updates and create the context just once?

Reply:
In general MSBuild operates on "configurations" which differs per global property set. If the global properties differ, MSBuild will in fact build the project multiple times. In fact, this is exactly how multitargetting works.

Project caching uses the same granularity as MSBuild uses for "build requests".

Reply:
Sorry for being slower here :-) The handler is still invoked for each single file access - and the `globalProperties` dictionary is created for each invocation. Is that correct?
I suppose the build request do not interleave - do they? Can we simply create the new 'build request associated data object' only if we do not have it some preexisting variable/field or if the build request identifier differs? This way we can balance between downsides of creating it for each single file access and caching all build request related data per all projects.

Or am I totally misunderstanding this?

Reply:
I created a cache